



```{r}
library(DeTREM)
library(Biobase)
library(dplyr)
library(biomaRt)
library(pheatmap)
library(Seurat)

```

```{r}

sc=readRDS("/restricted/projectnb/cteseq/data/spatial/genewhiz_20230323/sceFinal.rds")
bulk_expr_raw=readRDS("rounded_CTE_counts_for_deconv.rds")



#Changing ensgid to symbols
bulk_expr_raw$rownames <- rownames(bulk_expr_raw)
bulk_expr_raw$rownames <- gsub("\\.\\d+$", "", bulk_expr_raw$rownames)
#ensg_ids <- bulk_expr_raw$rownames
#ensembl.con <- useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl")
#gene_map <- getBM(attributes = c("ensembl_gene_id", "external_gene_name"),
#      filters = "ensembl_gene_id",
#      values = ensg_ids,
#      mart = ensembl.con)

#saveRDS(gene_map, file = "CTE_gene_map.rds")
gene_map = readRDS("CTE_gene_map.rds")

bulk_expr <- merge(bulk_expr_raw, gene_map, by.x = "rownames", by.y = "ensembl_gene_id", all.x = TRUE)
bulk_expr <- bulk_expr[!is.na(bulk_expr$external_gene_name), ]
bulk_expr <- bulk_expr[, !colnames(bulk_expr) %in% "rownames"]
bulk_expr <- subset(bulk_expr, external_gene_name != '')


bulk_expr <- aggregate(bulk_expr[ ,1:196], by = list(bulk_expr$external_gene_name), FUN = sum)

rownames(bulk_expr) <- bulk_expr$Group.1
bulk_expr <- bulk_expr[, !colnames(bulk_expr) %in% "Group.1"]
```


```{r}
# Extract the (gene x cell) count matrix as sc_expr. Alternative (gene x cell) count matrices can be loaded here
sc_expr=as.matrix(sc@assays@data@listData$counts)
sc_expr <- rowsum(sc_expr, row.names(sc_expr))
sc_expr_cte_and_rhi <- sc_expr[, grep("CTE|RHI", colnames(sc_expr)), drop = FALSE]


# Extract two relevant metadata columns from the seurat object, adjust variable names accordingly
# A data frame with celltype and sample columns can be loaded separately


celltype_column="CellType"
sample_column="sample"
sc_meta <- sc@colData[,c(celltype_column, sample_column)] %>% as.data.frame()
sc_meta <- sc_meta[sc_meta$CellType != "T Cells",]
sc_meta <- sc_meta[rownames(sc_meta) %in% colnames(sc_expr_cte_and_rhi),]
sc_expr_cte_and_rhi <- sc_expr_cte_and_rhi[,colnames(sc_expr_cte_and_rhi) %in% rownames(sc_meta)]
sc_meta=AnnotatedDataFrame(sc_meta)
#colnames(sc_meta)=c("CellType","sample")

# Generate the single cell (single nuclei) ExpressionSet object using counts and metadata
sc.eset=ExpressionSet(assayData=sc_expr_cte_and_rhi,phenoData=sc_meta)


# Load a bulk RNASeq count matrix, (gene x sample)
# We use an 'expected count' matrix generated by RSEM, but any count matrix should work

#using raw counts
bulk.eset=ExpressionSet(assayData=as.matrix(bulk_expr))

# Test the overlap of gene names between the two ExpressionSets
# Having a low number of 'TRUE' matches indicates a discrepancy and is a common error
table(rownames(bulk.eset) %in% rownames(sc.eset))
```

```{r}
# Run DeTREM using both ExpressionSets, extract cell-fraction estimates from the result (sample x celltype) matrix
estimates=DeTREM(bulk.eset = bulk.eset, sc.eset = sc.eset, clusters = 'CellType',samples = 'sample', verbose = F)
fractions=estimates$Est.prop.weighted
saveRDS(estimates, file = "DeTREM_CTE_estimates_all_no_t.rds")
```





```{r}
sc_expr=as.matrix(sc@assays@data@listData$counts)
sc_expr <- rowsum(sc_expr, row.names(sc_expr))

# Extract the (gene x cell) count matrix as sc_expr. Alternative (gene x cell) count matrices can be loaded here
#####indiv

sc_sample_names <- unique(sc$sample)
sc_sample_names <- sc_sample_names[! grepl("Control", sc_sample_names)]
sc_expr_indiv <- lapply(
  sc_sample_names,
  function(i) {
    sc_expr_indiv <- sc_expr[, grep(i, colnames(sc_expr)), drop = FALSE]
    return(sc_expr_indiv)
  })
names(sc_expr_indiv) <- sc_sample_names
#####

sc_expr_cte <- sc_expr[, grep("CTE", colnames(sc_expr)), drop = FALSE]
sc_expr_rhi <- sc_expr[, grep("RHI", colnames(sc_expr)), drop = FALSE]
#Control_subset <- sc_expr[, grep("Control", colnames(sc_expr)), drop = FALSE]


#CTE
#celltype_column <- sc@colData@listData$CellType
#sample_column <- sc@colData@listData$sample
sc_meta_cte <- sc@colData[,c("CellType", "sample")] %>% as.data.frame()
sc_meta_cte <- sc_meta_cte[sc_meta_cte$CellType != "T Cells",]
sc_meta_indiv_cte <- sc_meta_cte
sc_meta_cte <- sc_meta_cte[rownames(sc_meta_cte) %in% colnames(sc_expr_cte),]
sc_expr_cte_no_t <- sc_expr_cte[,colnames(sc_expr_cte) %in% rownames(sc_meta_cte)]
sc_meta_cte=AnnotatedDataFrame(sc_meta_cte)
####indiv
sc_meta_indiv <- lapply(
  sc_sample_names,
  function(i) {
    meta_indiv <- sc_meta_indiv_cte[rownames(sc_meta_indiv_cte) %in% colnames(sc_expr_indiv[[i]]),]
    return(meta_indiv)
  }
)
names(sc_meta_indiv) <- sc_sample_names

sc_expr_indiv_no_t <- lapply(
  sc_sample_names,
  function(i) {
    meta_indiv_no_t <- sc_expr_indiv[[i]][,colnames(sc_expr_indiv[[i]]) %in% rownames(sc_meta_indiv[[i]])]
    return(meta_indiv_no_t)
  }
)
names(sc_expr_indiv_no_t) <- sc_sample_names


for (i in sc_sample_names){
  sc_meta_indiv[[i]]=AnnotatedDataFrame(sc_meta_indiv[[i]])
}
####

#RHI

sc_meta_rhi <- sc@colData[,c("CellType", "sample")] %>% as.data.frame()
sc_meta_rhi <- sc_meta_rhi[sc_meta_rhi$CellType != "T Cells",]
sc_meta_rhi <- sc_meta_rhi[rownames(sc_meta_rhi) %in% colnames(sc_expr_rhi),]
sc_expr_rhi_no_t <- sc_expr_rhi[,colnames(sc_expr_rhi) %in% rownames(sc_meta_rhi)]
sc_meta_rhi=AnnotatedDataFrame(sc_meta_rhi)


# Generate the single cell (single nuclei) ExpressionSet object using counts and metadata
sc.eset_cte=ExpressionSet(assayData=sc_expr_cte_no_t,phenoData=sc_meta_cte)
sc.eset_rhi=ExpressionSet(assayData=sc_expr_rhi_no_t,phenoData=sc_meta_rhi)
####indiv
sc.eset_indiv <- lapply(
  sc_sample_names,
  function(i) {
    sc.eset=ExpressionSet(assayData=sc_expr_indiv_no_t[[i]],phenoData=sc_meta_indiv[[i]])
    return(sc.eset)
  }
)
names(sc.eset_indiv) <- sc_sample_names
####

# Load a bulk RNASeq count matrix, (gene x sample)
#using raw counts
bulk.eset=ExpressionSet(assayData=as.matrix(bulk_expr))

# Test the overlap of gene names between the two ExpressionSets
# Having a low number of 'TRUE' matches indicates a discrepancy and is a common error
table(rownames(bulk.eset) %in% rownames(sc.eset_cte))
table(rownames(bulk.eset) %in% rownames(sc.eset_rhi))
####indiv
for (i in sc_sample_names) {
  print(table(rownames(bulk.eset) %in% rownames(sc.eset_indiv[[i]])))
}
####
```

```{r}

# Run DeTREM using both ExpressionSets, extract cell-fraction estimates from the result (sample x celltype) matrix
estimates_cte=DeTREM(bulk.eset = bulk.eset, sc.eset = sc.eset_cte, clusters = 'CellType',samples = 'sample', verbose = F)
fractions_cte=estimates_cte$Est.prop.weighted
estimates_rhi=DeTREM(bulk.eset = bulk.eset, sc.eset = sc.eset_rhi, clusters = 'CellType',samples = 'sample', verbose = F)
fractions_rhi=estimates_rhi$Est.prop.weighted
saveRDS(estimates_cte, file = "DeTREM_CTE_estimates_cte.rds")
saveRDS(estimates_rhi, file = "DeTREM_CTE_estimates_rhi.rds")

```

indiv portion
```{r}
estimates_indiv <- lapply (
  sc_sample_names,
  function(i){
    indiv_est=DeTREM(bulk.eset = bulk.eset, sc.eset = na.omit(sc.eset_indiv[[i]]), clusters = 'CellType',samples = 'sample', verbose = F)
    return(indiv_est)
  }
)
names(estimates_indiv) <- sc_sample_names

###test
estimates_indiv_test=DeTREM(bulk.eset = bulk.eset, sc.eset = sc.eset_indiv$BM_CJ_k1049_CTE, clusters = 'CellType',samples = 'sample', verbose = F)
fractions_indiv_test=sc.eset_indiv$BM_CJ_k1049_CTE$Est.prop.weighted

###

fractions_indiv<- lapply(
  sc_sample_names,
  function(i) {
    frac_indiv <- estimates_indiv[[i]]$Est.prop.weighted
    return(frac_indiv)
  }
)
names(fractions_indiv) <- sc_sample_names

saveRDS(estimates_indiv, file = "DeTREM_CTE_estimates_indiv.rds")
```



