



```{r}
library(DeTREM)
library(Biobase)
library(dplyr)
library(biomaRt)
library(pheatmap)
library(Seurat)

```

```{r}

sc=readRDS("/restricted/projectnb/cteseq/data/spatial/genewhiz_20230323/sceFinal.rds")
bulk_expr_raw=readRDS("rounded_CTE_counts_for_deconv.rds")

#Changing ensgid to symbols
bulk_expr_raw$rownames <- rownames(bulk_expr_raw)
bulk_expr_raw$rownames <- gsub("\\.\\d+$", "", bulk_expr_raw$rownames)
ensg_ids <- bulk_expr_raw$rownames
ensembl.con <- useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl")
gene_map <- getBM(attributes = c("ensembl_gene_id", "external_gene_name"),
      filters = "ensembl_gene_id",
      values = ensg_ids,
      mart = ensembl.con)

#saveRDS(gene_map, file = "CTE_gene_map.rds")





bulk_expr <- merge(bulk_expr_raw, gene_map, by.x = "rownames", by.y = "ensembl_gene_id", all.x = TRUE)
bulk_expr <- bulk_expr[!is.na(bulk_expr$external_gene_name), ]
bulk_expr <- bulk_expr[, !colnames(bulk_expr) %in% "rownames"]
bulk_expr <- subset(bulk_expr, external_gene_name != '')


bulk_expr <- aggregate(bulk_expr[ ,1:196], by = list(bulk_expr$external_gene_name), FUN = sum)

rownames(bulk_expr) <- bulk_expr$Group.1
bulk_expr <- bulk_expr[, !colnames(bulk_expr) %in% "Group.1"]
```


```{r}
# Extract the (gene x cell) count matrix as sc_expr. Alternative (gene x cell) count matrices can be loaded here
sc_expr=as.matrix(sc@assays@data@listData$counts)
sc_expr <- rowsum(sc_expr, row.names(sc_expr))


# Extract two relevant metadata columns from the seurat object, adjust variable names accordingly
# A data frame with celltype and sample columns can be loaded separately
#celltype_column=sc@colData@listData$CellType
#sample_column=sc@colData@listData$sample
#sc_meta=data.frame(CellType = celltype_column, sample = sample_column)

celltype_column="CellType"
sample_column="sample"
sc_meta <- sc@colData[,c(celltype_column, sample_column)] %>% as.data.frame()
sc_meta=AnnotatedDataFrame(sc_meta)
colnames(sc_meta)=c("CellType","sample")


# Generate the single cell (single nuclei) ExpressionSet object using counts and metadata
sc.eset=ExpressionSet(assayData=sc_expr,phenoData=sc_meta)



# Load a bulk RNASeq count matrix, (gene x sample)
# We use an 'expected count' matrix generated by RSEM, but any count matrix should work

#using raw counts
bulk.eset=ExpressionSet(assayData=as.matrix(bulk_expr))

# Test the overlap of gene names between the two ExpressionSets
# Having a low number of 'TRUE' matches indicates a discrepancy and is a common error
table(rownames(bulk.eset) %in% rownames(sc.eset))
```

```{r}
# Run DeTREM using both ExpressionSets, extract cell-fraction estimates from the result (sample x celltype) matrix
estimates=DeTREM(bulk.eset = bulk.eset, sc.eset = sc.eset, clusters = 'CellType',samples = 'sample', verbose = F)
fractions=estimates$Est.prop.weighted
saveRDS(estimates, file = "DeTREM_CTE_estimates_all.rds")
```


Striations
```{r}
#unique(sc@colData@listData$PathologicalGroup)

#CTE_subset <- sc[, colData(sc)$PathologicalGroup == "CTE"]
#RHI_subset <- sc[, colData(sc)$PathologicalGroup == "RHI"]
#Control_subset <- sc[, colData(sc)$PathologicalGroup == "Control"]

CTE_subset <- sc_expr[, grep("CTE", colnames(sc_expr)), drop = FALSE]
RHI_subset <- sc_expr[, grep("RHI", colnames(sc_expr)), drop = FALSE]
Control_subset <- sc_expr[, grep("Control", colnames(sc_expr)), drop = FALSE]

```



```{r}
# Extract the (gene x cell) count matrix as sc_expr. Alternative (gene x cell) count matrices can be loaded here
#sc_expr_cte=as.matrix(CTE_subset@assays@data@listData$counts)
#sc_expr_rhi=as.matrix(RHI_subset@assays@data@listData$counts)
#sc_expr_ctrl=as.matrix(Control_subset@assays@data@listData$counts)

#saveRDS(sc_expr_cte, file = "sc_cte_expr.rds")
#saveRDS(sc_expr_rhi, file = "sc_rhi_expr.rds")
#saveRDS(sc_expr_ctrl, file = "sc_ctrl_expr.rds")

#sc_expr_cte <- readRDS("sc_cte_expr.rds")
#sc_expr_rhi <- readRDS("sc_rhi_expr.rds")
#sc_expr_ctrl <- readRDS("sc_ctrl_expr.rds")


#sc_expr_cte <- rowsum(sc_expr_cte, row.names(sc_expr_cte))
#sc_expr_rhi <- rowsum(sc_expr_rhi, row.names(sc_expr_rhi))
#sc_expr_ctrl <- rowsum(sc_expr_ctrl, row.names(sc_expr_ctrl))







# Extract two relevant metadata columns from the seurat object, adjust variable names accordingly
# A data frame with celltype and sample columns can be loaded separately
#celltype_column=sc@colData@listData$CellType
#sample_column=sc@colData@listData$sample
#sc_meta=data.frame(CellType = celltype_column, sample = sample_column)

celltype_column="CellType"
sample_column="sample"
sc_meta_cte <- CTE_subset@colData[,c(celltype_column, sample_column)] %>% as.data.frame()
sc_meta_rhi <- RHI_subset@colData[,c(celltype_column, sample_column)] %>% as.data.frame()
sc_meta_ctrl <- Control_subset@colData[,c(celltype_column, sample_column)] %>% as.data.frame()

sc_meta_cte=AnnotatedDataFrame(sc_meta_cte)
sc_meta_rhi=AnnotatedDataFrame(sc_meta_rhi)
sc_meta_ctrl=AnnotatedDataFrame(sc_meta_ctrl)

colnames(sc_meta_cte)=c("CellType","sample")
colnames(sc_meta_rhi)=c("CellType","sample")
colnames(sc_meta_ctrl)=c("CellType","sample")


# Generate the single cell (single nuclei) ExpressionSet object using counts and metadata
sc.eset_cte=ExpressionSet(assayData=sc_expr_cte,phenoData=sc_meta_cte)
sc.eset_rhi=ExpressionSet(assayData=sc_expr_rhi,phenoData=sc_meta_rhi)
sc.eset_ctrl=ExpressionSet(assayData=sc_expr_ctrl,phenoData=sc_meta_ctrl)



# Load a bulk RNASeq count matrix, (gene x sample)
# We use an 'expected count' matrix generated by RSEM, but any count matrix should work

#using raw counts
bulk.eset=ExpressionSet(assayData=as.matrix(bulk_expr))

# Test the overlap of gene names between the two ExpressionSets
# Having a low number of 'TRUE' matches indicates a discrepancy and is a common error
table(rownames(bulk.eset) %in% rownames(sc.eset_cte))
table(rownames(bulk.eset) %in% rownames(sc.eset_rhi))
table(rownames(bulk.eset) %in% rownames(sc.eset_ctrl))


```

```{r}

# Run DeTREM using both ExpressionSets, extract cell-fraction estimates from the result (sample x celltype) matrix
estimates=DeTREM(bulk.eset = bulk.eset, sc.eset = sc.eset, clusters = 'CellType',samples = 'sample', verbose = F)
fractions=estimates$Est.prop.weighted
saveRDS(estimates, file = "DeTREM_CTE_estimates_all.rds")

```




